apps/admin_api/lib/admin_api/v1/controllers/tranasction_calculation_controller.ex

```diff
@@ -0,0 +1,57 @@
+defmodule AdminAPI.V1.TransactionCalculationController do
+  @moduledoc """
+  The controller to serve transaction calculations.
+  """
+  use AdminAPI, :controller
+  import AdminAPI.V1.ErrorHandler
+  alias EWallet.Exchange
+  alias EWalletDB.Token
+
+  @doc """
+  Calculates transaction amounts.
+  """
+  def calculate(conn, attrs) do
+    from_amount = attrs["from_amount"]
+    from_token = Token.get(attrs["from_token_id"])
+    to_amount = attrs["to_amount"]
+    to_token = Token.get(attrs["to_token_id"])
+
+    case do_calculate(from_amount, from_token, to_amount, to_token) do
+      {:ok, calculation} ->
+        render(conn, :calculation, %{calculation: calculation})
+
+      {:error, code, description} ->
+        handle_error(conn, code, description)
+    end
+  end
+
+  defp do_calculate(from_amount, from_token_id, to_amount, to_token_id)
+
+  defp do_calculate(nil, _, nil, _) do
+    {:error, :invalid_parameter, "either `from_amount` or `to_amount` is required"}
+  end
+
+  defp do_calculate(_, nil, _, nil) do
+    {:error, :invalid_parameter, "both `from_token_id` and `to_token_id` are required"}
+  end
+
+  defp do_calculate(_, nil, _, _) do
+    {:error, :invalid_parameter, "`from_token_id` is required"}
+  end
+
+  defp do_calculate(_, _, _, nil) do
+    {:error, :invalid_parameter, "`to_token_id` is required"}
+  end
+
+  defp do_calculate(nil, from_token, to_amount, to_token) do
+    Exchange.calculate(nil, from_token, to_amount, to_token)
+  end
+
+  defp do_calculate(from_amount, from_token, nil, to_token) do
+    Exchange.calculate(from_amount, from_token, nil, to_token)
+  end
+
+  defp do_calculate(from_amount, from_token, to_amount, to_token) do
```

Comment:
Can't the two functions above be removed? Since they are just passing the same value for the same argument.

Reply:
The two functions above go to `Exchange.calculate/4`, hence either allowed a nil from_amount or to_amount but not both. While the last one goes to `Exchange.validate/4`.

Could've done those two in one function with guards but it will look like this which looks confusing to me:

```ex
  defp do_calculate(from_amount, from_token, to_amount, to_token)
       when is_nil(from_amount) or is_nil(to_amount) do
    Exchange.calculate(nil, from_token, to_amount, to_token)
  end
```

I'll leave this one as is for this PR but open to more suggestions.